File: ios/VisionMLModule.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(VisionML, NSObject)

// Object-based detector API
RCT_EXTERN_METHOD(createDetector:(NSString *)modelPath
                  classLabels:(NSArray *)classLabels
                  inputSize:(NSNumber *)inputSize
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(detect:(NSString *)detectorId
                  imageUri:(NSString *)imageUri
                  confidenceThreshold:(NSNumber *)confidenceThreshold
                  iouThreshold:(NSNumber *)iouThreshold
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(disposeDetector:(NSString *)detectorId
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(disposeAllDetectors:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

// Live Activity Methods
RCT_EXTERN_METHOD(isLiveActivityAvailable:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(startVideoScanActivity:(NSString *)videoName
                  videoDuration:(NSNumber *)videoDuration
                  scanMode:(NSString *)scanMode
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(updateVideoScanActivity:(NSNumber *)progress
                  phase:(NSString *)phase
                  nsfwCount:(NSNumber *)nsfwCount
                  framesAnalyzed:(NSNumber *)framesAnalyzed
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(endVideoScanActivity:(NSNumber *)nsfwCount
                  framesAnalyzed:(NSNumber *)framesAnalyzed
                  isNSFW:(BOOL)isNSFW
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

// Video Analysis Methods
RCT_EXTERN_METHOD(analyzeVideo:(NSString *)detectorId
                  assetId:(NSString *)assetId
                  mode:(NSString *)mode
                  sampleInterval:(NSNumber *)sampleInterval
                  confidenceThreshold:(NSNumber *)confidenceThreshold
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(quickCheckVideo:(NSString *)detectorId
                  assetId:(NSString *)assetId
                  confidenceThreshold:(NSNumber *)confidenceThreshold
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

// Vision Framework Methods
RCT_EXTERN_METHOD(analyzeAnimals:(NSString *)assetId
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(analyzeHumanPose:(NSString *)assetId
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(analyzeComprehensive:(NSString *)assetId
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

// Sensitive Content Analysis Methods (iOS 17+)
RCT_EXTERN_METHOD(getSensitiveContentAnalysisStatus:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(openSensitiveContentSettings:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(analyzeSensitiveContent:(NSString *)assetId
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(batchAnalyzeSensitiveContent:(NSArray *)assetIds
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

// SCA Video Analysis Methods (iOS 17+)
RCT_EXTERN_METHOD(analyzeVideoSensitiveContent:(NSString *)assetId
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(batchAnalyzeVideosSensitiveContent:(NSArray *)assetIds
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

+ (BOOL)requiresMainQueueSetup {
  return NO;
}

@end

File: android/src/main/java/com/visionml/VisionMLModule.kt
package com.visionml

import com.facebook.react.bridge.*

/**
 * React Native module for VisionML
 *
 * TODO: Implement Android support
 * See ANDROID_PLAN.md for implementation details
 */
class VisionMLModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private val detectors = mutableMapOf<String, ONNXInference>()
    private var detectorIdCounter = 0

    override fun getName(): String = "VisionML"

    // MARK: - Detector Management

    @ReactMethod
    fun createDetector(
        modelPath: String,
        classLabels: ReadableArray,
        inputSize: Int,
        promise: Promise
    ) {
        // TODO: Implement
        promise.reject("NOT_IMPLEMENTED", "Android support not yet implemented")
    }

    @ReactMethod
    fun detect(
        detectorId: String,
        imageUri: String,
        confidenceThreshold: Double,
        iouThreshold: Double,
        promise: Promise
    ) {
        // TODO: Implement
        promise.reject("NOT_IMPLEMENTED", "Android support not yet implemented")
    }

    @ReactMethod
    fun disposeDetector(detectorId: String, promise: Promise) {
        // TODO: Implement
        promise.reject("NOT_IMPLEMENTED", "Android support not yet implemented")
    }

    @ReactMethod
    fun disposeAllDetectors(promise: Promise) {
        // TODO: Implement
        promise.reject("NOT_IMPLEMENTED", "Android support not yet implemented")
    }

    // MARK: - Video Analysis

    @ReactMethod
    fun analyzeVideo(
        detectorId: String,
        assetId: String,
        mode: String,
        sampleInterval: Double,
        confidenceThreshold: Double,
        promise: Promise
    ) {
        // TODO: Implement using MediaMetadataRetriever
        promise.reject("NOT_IMPLEMENTED", "Android support not yet implemented")
    }

    @ReactMethod
    fun quickCheckVideo(
        detectorId: String,
        assetId: String,
        confidenceThreshold: Double,
        promise: Promise
    ) {
        // TODO: Implement
        promise.reject("NOT_IMPLEMENTED", "Android support not yet implemented")
    }

    // MARK: - ML Kit Analysis (Vision Framework equivalent)

    @ReactMethod
    fun analyzeAnimals(assetId: String, promise: Promise) {
        // TODO: Implement using ML Kit Image Labeling
        promise.reject("NOT_IMPLEMENTED", "Android support not yet implemented")
    }

    @ReactMethod
    fun analyzeHumanPose(assetId: String, promise: Promise) {
        // TODO: Implement using ML Kit Pose Detection
        promise.reject("NOT_IMPLEMENTED", "Android support not yet implemented")
    }

    @ReactMethod
    fun analyzeComprehensive(assetId: String, promise: Promise) {
        // TODO: Implement using ML Kit (multiple detectors)
        promise.reject("NOT_IMPLEMENTED", "Android support not yet implemented")
    }

    // MARK: - Progress Notification (Live Activity equivalent)

    @ReactMethod
    fun isLiveActivityAvailable(promise: Promise) {
        // Android always returns true - we use foreground notification
        promise.resolve(true)
    }

    @ReactMethod
    fun startVideoScanActivity(
        videoName: String,
        videoDuration: Double,
        scanMode: String,
        promise: Promise
    ) {
        // TODO: Start foreground service with notification
        promise.resolve(
            Arguments.createMap().apply {
                putNull("activityId")
                putBoolean("success", false)
            }
        )
    }

    @ReactMethod
    fun updateVideoScanActivity(
        progress: Double,
        phase: String,
        nsfwCount: Int,
        framesAnalyzed: Int,
        promise: Promise
    ) {
        // TODO: Update foreground notification
        promise.resolve(false)
    }

    @ReactMethod
    fun endVideoScanActivity(
        nsfwCount: Int,
        framesAnalyzed: Int,
        isNSFW: Boolean,
        promise: Promise
    ) {
        // TODO: End foreground service
        promise.resolve(false)
    }
}
